modules/api/src/main/java/org/apache/ignite/lang/CancelHandleImpl.java

```diff
@@ -0,0 +1,149 @@
+
+/** Implementation of {@link CancelHandle}. */
+final class CancelHandleImpl implements CancelHandle {
+
+    private final CompletableFuture<Void> cancelFut = new CompletableFuture<>();
+
+    private final CancellationTokenImpl token;
+
+    CancelHandleImpl() {
+        this.token = new CancellationTokenImpl(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void cancel() {
+        doCancelAsync();
+
+        cancelFut.join();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Void> cancelAsync() {
+        doCancelAsync();
+
+        return cancelFut;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isCancelled() {
+        return token.isCancelled();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CancellationToken token() {
+        return token;
+    }
+
+    private void doCancelAsync() {
+        token.cancel();
+    }
+
+    static final class CancellationTokenImpl implements CancellationToken {
+
+        private final ArrayDeque<Cancellation> cancellations = new ArrayDeque<>();
+
+        private final CancelHandleImpl handle;
+
+        private final Object mux = new Object();
+
+        private volatile CompletableFuture<Void> cancelFut;
+
+        CancellationTokenImpl(CancelHandleImpl handle) {
+            this.handle = handle;
+        }
+
+        void addCancelAction(Runnable action, CompletableFuture<?> fut) {
+            Cancellation cancellation = new Cancellation(action, fut);
+
+            if (cancelFut != null) {
+                cancellation.run();
+            } else {
+                synchronized (mux) {
+                    if (cancelFut == null) {
+                        cancellations.add(cancellation);
+                    }
+                }
+            }
+        }
+
+        CompletableFuture<Void> cancelHandleFut() {
+            return handle.cancelFut;
+        }
+
+        boolean isCancelled() {
+            return cancelFut != null;
+        }
+
+        @SuppressWarnings("rawtypes")
+        void cancel() {
+            if (cancelFut != null) {
+                return;
+            }
+
+            synchronized (mux) {
+                if (cancelFut != null) {
+                    return;
+                }
+
+                // First assemble all completion futures
+                CompletableFuture[] futures = cancellations.stream()
+                        .map(c -> c.completionFut)
+                        .toArray(CompletableFuture[]::new);
+
+                // handle.cancelFut completes when all cancellation futures complete.
+                cancelFut = CompletableFuture.allOf(futures).whenComplete((r, t) -> {
+                    handle.cancelFut.complete(null);
+                });
+            }
+
+            // Run cancellation actions outside of lock
+            for (Cancellation cancellation : cancellations) {
+                cancellation.run();
+            }
```

Comment:
we need to introduce error handling here: catch exceptions, compose a single one via `addSuppressed` and, perhaps, complete `handle.cancelFut` exceptionally (this must be reflected in javadoc).

Reply:
Fixed.